### ISSUE TITLE
`cephadm bootstap` hangs in `mgr dump`

### ISSUE THREAD

**Issue by sebastian-philipp** [ceph/ceph-salt#130]
Using https://github.com/ceph/ceph/pull/34031 I was able to run `cephadm boostrap` **successfully** locally:

```
/usr/bin/python3 /usr/sbin/cephadm --image registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph --verbose bootstrap --mon-ip 10.20.62.200 --config /root/ceph.conf --initial-dashboard-user admin --output-keyring /etc/ceph/ceph.client.admin.keyring --output-config /etc/ceph/ceph.conf --skip-prepare-host --skip-pull --skip-ssh
```

Running this command via salt hangs:

```
salt 'master.octopus.com' cmd.run '/usr/bin/python3 /usr/sbin/cephadm --image registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph --verbose bootstrap --mon-ip 10.20.62.200 --config /root/ceph.conf --initial-dashboard-user admin --output-keyring /etc/ceph/ceph.client.admin.keyring --output-config /etc/ceph/ceph.conf --skip-prepare-host --skip-pull --skip-ssh'
```

```
master:~ # pstree -a 
systemd --switched-root --system --deserialize 23
  ├─bash /var/lib/ceph/29ce8ec0-6936-11ea-aa3f-5254009bd4ee/mon.master/unit.run
  │   └─podman run --rm --net=host --privileged --group-add=disk --name ceph-29ce8ec0-6936-11ea-aa3f-5254009bd4ee-mon.master -e CONTAINER_IMAGE=registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph -e NODE_NAME=master -v/var/run/ce
  │       └─10*[{podman}]
  ├─bash /var/lib/ceph/29ce8ec0-6936-11ea-aa3f-5254009bd4ee/mgr.master.mmzyez/unit.run
  │   └─podman run --rm --net=host --name ceph-29ce8ec0-6936-11ea-aa3f-5254009bd4ee-mgr.master.mmzyez -e CONTAINER_IMAGE=registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph -e NODE_NAME=master -v/var/run/ceph/29ce8ec0-6936-11ea-a
  │       └─9*[{podman}]
  ├─chronyd
  ├─conmon -s -c e82a11f65ab4953bd7eb5b7c4e06f47698bcc0ab8a50819490b63920e0d1b1a4 -u e82a11f65ab4953bd7eb5b7c4e06f47698bcc0ab8a50819490b63920e0d1b1a4 -n ceph-29ce8ec0-6936-11ea-aa3f-5254009bd4ee-mon.master -r /usr/bin/runc -b/var/lib/c
  │   ├─ceph-mon -n mon.master -f --setuser ceph --setgroup ceph --default-log-to-file=false --default-log-to-stderr=true --default-log-stderr-prefix=debug  --default-mon-cluster-log-to-file=false --default-mon-cluster-log-to-stderr=true
  │   │   └─25*[{ceph-mon}]
  │   └─{conmon}
  ├─conmon -s -c 860596f4e4ebe06a80130256372d79aba6638a43851654169381a794c62ca680 -u 860596f4e4ebe06a80130256372d79aba6638a43851654169381a794c62ca680 -n ceph-29ce8ec0-6936-11ea-aa3f-5254009bd4ee-mgr.master.mmzyez -r /usr/bin/runc -b/va
  │   ├─ceph-mgr -n mgr.master.mmzyez -f --setuser ceph --setgroup ceph --default-log-to-file=false --default-log-to-stderr=true --default-log-stderr-prefix=debug 
  │   │   └─62*[{ceph-mgr}]
  │   └─{conmon}
  ├─conmon -s -c 446b1521faf269b116f0aed4500cae54dadf25687ba868cc19280d487581de36 -u 446b1521faf269b116f0aed4500cae54dadf25687ba868cc19280d487581de36 -n elastic_cerf -r /usr/bin/runc -b/var/lib/containers/storage/overlay-containers/446
  │   ├─(ceph)
  │   └─{conmon}
  ├─salt-master /usr/bin/salt-master
  │   ├─salt-master /usr/bin/salt-master
  │   ├─salt-master /usr/bin/salt-master
  │   │   └─2*[{salt-master}]
  │   ├─salt-master /usr/bin/salt-master
  │   ├─salt-master /usr/bin/salt-master
  │   ├─salt-master /usr/bin/salt-master
  │   │   ├─salt-master /usr/bin/salt-master
  │   │   │   └─6*[{salt-master}]
  │   │   ├─salt-master /usr/bin/salt-master
  │   │   │   └─4*[{salt-master}]
  │   │   ├─salt-master /usr/bin/salt-master
  │   │   │   └─4*[{salt-master}]
  │   │   ├─salt-master /usr/bin/salt-master
  │   │   │   └─4*[{salt-master}]
  │   │   ├─salt-master /usr/bin/salt-master
  │   │   │   └─4*[{salt-master}]
  │   │   └─salt-master /usr/bin/salt-master
  │   │       └─4*[{salt-master}]
  │   └─salt-master /usr/bin/salt-master
  │       └─{salt-master}
  ├─salt-minion /usr/bin/salt-minion
  │   └─salt-minion /usr/bin/salt-minion
  │       ├─salt-minion /usr/bin/salt-minion
  │       ├─salt-minion /usr/bin/salt-minion
  │       │   └─python3 /usr/sbin/cephadm --image registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph --verbose bootstrap --mon-ip 10.20.62.200 --config /root/ceph.conf --initial-dashboard-user admin --output-keyring...
  │       │       └─/usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph -e NODE_NAME=master -v /var/log/ceph/29ce8ec0-6936-11ea-aa3f-5254009bd4ee:/var/log/ceph:z -v /tmp/ceph-tmp74qa4op2:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmp0n3xbawl:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph mgr dump
  │       │           ├─runc start 446b1521faf269b116f0aed4500cae54dadf25687ba868cc19280d487581de36
  │       │           │   └─6*[{runc}]
  │       │           └─10*[{podman}]
  │       └─3*[{salt-minion}]
```

**Comment by bk201** [comment]
>   ├─conmon -s -c 446b1521faf269b116f0aed4500cae54dadf25687ba868cc19280d487581de36 -u 446b1521faf269b116f0aed4500cae54dadf25687ba868cc19280d487581de36 -n elastic_cerf -r /usr/bin/runc -b/var/lib/containers/storage/overlay-containers/446
>   │   ├─(ceph)    <-- Zombie
>   │   └─{conmon}

The `(ceph)` (enclosed with `( )`) is a zombie process.


**tl;dr:** Running `ceph mgr dump` inside a container with salt-minion's remote run can reproduce the issue. Maybe it's not related to cephadm script.

Some experiments:

* Bootstrap and let it hang
* Find the zombie process
```
master:~ # ps aux | grep Z
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root     17649  0.1  0.0      0     0 ?        Zs   07:37   0:00 [ceph] <defunct>
root     18339  0.0  0.0   8680   900 pts/1    R+   07:43   0:00 grep Z

master:~ # pstree -apl -s 17649
systemd,1 --switched-root --system --deserialize 22
  └─conmon,17614 -s -c a42aff2600e62260b9720da5b0c454c829dc02279fd5897b41602305ec75eec2 -u a42aff2600e62260b9720da5b0c454c829dc02279fd5897b41602305ec75eec2 -n quirky_gates -r /usr/bin/runc -b /var/lib/containers/storage/overlay-containers/a42aff2600e62260b9720da5b0c454c829dc02279fd5897b41602305ec75eec2/userdata -p /var/run/containers/storage/overlay-containers/a42aff2600e62260b9720da5b0c454c829dc02279fd5897b41602305ec75eec2/userdata/pidfile --exit-dir /var/run/libpod/exits --runtime-arg --log-format=json --runtime-arg --log --runtime-arg=/var/run/containers/storage/overlay-containers/a42aff2600e62260b9720da5b0c454c829dc02279fd5897b41602305ec75eec2/userdata/oci-log --exit-command /usr/bin/podman --exit-command-arg --root --exit-command-arg /var/lib/containers/storage --exit-command-arg --runroot --exit-command-arg /var/run/containers/storage --exit-command-arg --log-level --exit-command-arg error --exit-command-arg --cgroup-manager --exit-command-arg systemd --exit-command-arg --tmpdir --exit-command-arg /var/run/libpod --exit-command-arg --runtime --exit-command-arg runc --exit-command-arg container --exit-command-arg cleanup --exit-command-arg --rm --exit-command-arg a42aff2600e62260b9720da5b0c454c829dc02279fd5897b41602305ec75eec2 --socket-dir-path /var/run/libpod/socket -l k8s-file:/var/lib/containers/storage/overlay-containers/a42aff2600e62260b9720da5b0c454c829dc02279fd5897b41602305ec75eec2/userdata/ctr.log --log-level error
      └─(ceph,17649)
```
* kill parent `conmon` process, `podman` command can work again. Container `a42aff2600e6` is the one runs `ceph mgr dump`.
```
master:~ # kill -9 17614
master:~ # podman ps
CONTAINER ID  IMAGE                                     COMMAND               CREATED         STATUS             PORTS  NAMES
a42aff2600e6  172.17.0.1:5000/ceph-ci/ceph:octopus-wip  /usr/bin/ceph mgr...  11 minutes ago  Up 1 second ago           quirky_gates
4e16d276fd7d  172.17.0.1:5000/ceph-ci/ceph:octopus-wip  /usr/bin/ceph-mgr...  11 minutes ago  Up 11 minutes ago         ceph-05c15c1c-69ac-11ea-af38-52540098c2d4-mgr.master.xztwpb
e1737f9e77ac  172.17.0.1:5000/ceph-ci/ceph:octopus-wip  /usr/bin/ceph-mon...  11 minutes ago  Up 11 minutes ago         ceph-05c15c1c-69ac-11ea-af38-52540098c2d4-mon.master
```

Another attempt is to run `ceph mgr dump` with or without salt minon:
* Without salt minion, works and output the very long JSON.
```
master:~ # /usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=172.17.0.1:5000/ceph-ci/ceph:octopus-wip -e NODE_NAME=master -v /var/log/ceph/05c15c1c-69ac-11ea-af38-52540098c2d4:/var/log/ceph:z -v /tmp/ceph-tmp_013m2pg:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmpq_t3oisy:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph 172.17.0.1:5000/ceph-ci/ceph:octopus-wip mgr dump

(...very long output...)
```
* WIth salt minion, hang again. And `podman ps` command no longer work. Another zombie process is created.
```
master:~ # salt 'master*' cmd.run "/usr/bin/podman run --rm --net=host -e CONTAINER_IMAGE=172.17.0.1:5000/ceph-ci/ceph:octopus-wip -e NODE_NAME=master -v /var/log/ceph/05c15c1c-69ac-11ea-af38-52540098c2d4:/var/log/ceph:z -v /tmp/ceph-tmp_013m2pg:/etc/ceph/ceph.client.admin.keyring:z -v /tmp/ceph-tmpq_t3oisy:/etc/ceph/ceph.conf:z --entrypoint /usr/bin/ceph 172.17.0.1:5000/ceph-ci/ceph:octopus-wip mgr dump"

( hang again)

master:~ # ps aux | grep Z
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root     19128  1.0  0.0      0     0 ?        Zs   07:52   0:00 [ceph] <defunct>
root     19235  0.0  0.0   8680   812 pts/2    S+   07:52   0:00 grep Z
```

**Comment by sebastian-philipp** [comment]
yes, interesting to see, that this is independent of ceph-salt and simply caused by `cmd.run`

**Comment by ricardoasmarques** [comment]
Some more information I've collected:
```
master:~ # salt node1\* cmd.run_all 'cephadm shell -- ceph mgr dump'
<HANGS>

master:~ # salt node1\* cmd.run_all 'ceph mgr dump'
<WORKS>

master:~ # ssh node1
node1:~ # salt-call cmd.run_all 'cephadm shell -- ceph mgr dump'
<WORKS>
```

**Comment by sebastian-philipp** [comment]
so far, we've seen:

* `cephadm` is innocent (`salt 'master*' cmd.run "/usr/bin/podman ..`. hangs)
* we need podman in the chain (`salt node1\* cmd.run_all 'ceph mgr dump'` works)

let's see if we can reduce the example further. 

```
master:~ # seq 10000  | awk '{print 12345678901234567890}' | wc
  10000   10000  210000
```
next:
```
master:~ # salt '*' cmd.run "seq 10000  | awk '{print 12345678901234567890}'"
<works>
```
next:
```
master:~ # salt '*' cmd.run 'podman run --net=host --rm registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph bash -c "seq 10000  | awk '"'"'{print 12345678901234567890}'"'"' "'
^C
<hangs>
```

next (1000 instead of 10000):
```
salt '*' cmd.run 'podman run --net=host --rm registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph bash -c "seq 1000  | awk '"'"'{print 12345678901234567890}'"'"' "'
<works>
```

now we can actually look where it hangs:

```
salt '*' cmd.run 'podman run --net=host --rm registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph bash -c "seq 5070  | awk '"'"'{print 12345678901234567890}'"'"' "'
<works>
```
but:
```
salt '*' cmd.run 'podman run --net=host --rm registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph bash -c "seq 5080  | awk '"'"'{print 12345678901234567890}'"'"' "'
^C
<hangs>
```

which is roughly about 105000 bytes. interesting.

also hangs with other images:
```
salt '*' cmd.run 'podman run --net=host --rm registry.opensuse.org/opensuse/leap:15.2 bash -c "seq 5080 | awk '"'"'{print 12345678901234567890}'"'"' "'
```

salt-call works:
```
salt-call cmd.run_all 'podman run --net=host --rm registry.opensuse.org/opensuse/leap:15.2 bash -c "seq 5080 | awk '"'"'{print 12345678901234567890}'"'"'  "'
```

**Comment by ricardoasmarques** [comment]
And it also hangs if we try to `hide_output` [1]

```
salt '*' cmd.run 'podman run --net=host --rm registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph bash -c "seq 5080  | awk '"'"'{print 12345678901234567890}'"'"' "' hide_output=True
<HANGS>
```

[1] https://github.com/saltstack/salt/blob/master/salt/modules/cmdmod.py#L1069

**Comment by ricardoasmarques** [comment]
I've done the same test using **docker** and it works:

```
salt 'node1*' cmd.run 'docker run --net=host --rm registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph bash -c "seq 5080  | awk '"'"'{print 12345678901234567890}'"'"' "'
<WORKS>
```

If I try with very large output, it still returns without hanging:

```
master:~ # salt 'node1*' cmd.run 'docker run --net=host --rm registry.opensuse.org/filesystems/ceph/octopus/upstream/images/ceph/ceph bash -c "seq 95080  | awk '"'"'{print 12345678901234567890}'"'"' "'
node1.octopus.com:
    VALUE_TRIMMED
```

**Comment by ricardoasmarques** [comment]
I've used podman `1.4.4` for my tests:
```
node1:~ # podman --version
podman version 1.4.4
```

**Comment by bk201** [comment]
Confirm it's only reproduced when run via salt-minion and not related to Ceph.

Using gdb to attach the parent of zombie process:
```
(gdb) bt
#0  0x00007f5919f2c23b in write () from /lib64/libc.so.6
#1  0x0000000000405b2c in write_all ()
#2  0x0000000000407a0f in read_stdio ()
#3  0x0000000000407af0 in stdio_cb ()
#4  0x00007f591a24c3b4 in g_main_context_dispatch () from /usr/lib64/libglib-2.0.so.0
#5  0x00007f591a24c750 in ?? () from /usr/lib64/libglib-2.0.so.0
#6  0x00007f591a24ca12 in g_main_loop_run () from /usr/lib64/libglib-2.0.so.0
#7  0x00000000004055e2 in main ()
```
https://github.com/containers/conmon/blob/master/src/conmon.c#L377
Still no clues why its child becomes zombie.

```
podman version 1.8.1
conmon version: master branch
```

**Comment by tserong** [comment]
Also confirmed with podman 1.8.0 and conmon 2.0.10 from Devel:Storage:7:

This works:
`salt master.ses7-mini.com cmd.run 'bash -c "for n in \$(seq 5070) ; do /bin/echo 12345678901234567890 ; done"'`

This hangs:
`salt master.ses7-mini.com cmd.run 'podman run --net=host --rm registry.suse.de/devel/storage/7.0/cr/containers/ses/7/ceph/ceph bash -c "for n in \$(seq 5070) ; do /bin/echo 12345678901234567890 ; done"'`

This works:
`podman run --net=host --rm registry.suse.de/devel/storage/7.0/cr/containers/ses/7/ceph/ceph bash -c 'for n in $(seq 5070) ; do /bin/echo 12345678901234567890 ; done'`

(I didn't use the awk variant from previous comments, because apparently the ceph container image doesn't include the awk binary?)

**Comment by sebastian-philipp** [comment]
> This hangs:
> `salt master.ses7-mini.com cmd.run 'podman run --net=host --rm registry.suse.de/devel/storage/7.0/cr/containers/ses/7/ceph/ceph bash -c "for n in \$(seq 5070) ; do /bin/echo 12345678901234567890 ; done"'`

Interesting, as it is less than 106 kB! (`seq 5080`)In my tests, this always worked.

> (I didn't use the awk variant from previous comments, because apparently the ceph container image doesn't include the awk binary?)

ok, this is awkward.

**Comment by bk201** [comment]
I checked `quay.io/ceph-ci/ceph:octopus` and `registry.opensuse.org/filesystems/ceph/master/upstream/images/ceph/ceph`. They should contain awk binary.
> (I didn't use the awk variant from previous comments, because apparently the ceph container image doesn't include the awk binary?)

**Comment by sebastian-philipp** [comment]
Possible workaround:

> ```ini
> master:~ # cat /usr/lib/systemd/system/salt-minion.service
> [Unit]
> Description=The Salt Minion
> After=network.target salt-master.service
> 
> [Service]
> #Type=notify
> #NotifyAccess=all
> LimitNOFILE=8192
> ExecStart=/usr/bin/salt-minion
> KillMode=process
> Restart=on-failure
> RestartSec=15
> 
> [Install]
> WantedBy=multi-user.target
> ```

**Comment by tserong** [comment]
> I checked `quay.io/ceph-ci/ceph:octopus` and `registry.opensuse.org/filesystems/ceph/master/upstream/images/ceph/ceph`. They should contain awk binary.
> 
> > (I didn't use the awk variant from previous comments, because apparently the ceph container image doesn't include the awk binary?)

I'm using registry.suse.de/devel/storage/7.0/cr/containers/ses/7/ceph/ceph, which AFAICT doesn't include awk

**Comment by smithfarm** [comment]
I wonder if there's something about the runc patch in https://github.com/opencontainers/runc/pull/1807 which makes it work only up to a certain number of characters in the command output.

**Comment by sebastian-philipp** [comment]
might relate to https://github.com/containers/libpod/issues/4600

**Comment by sebastian-philipp** [comment]
```
salt '*' cmd.run 'unset NOTIFY_SOCKET; podman run --net=host --rm registry.opensuse.org/opensuse/leap:15.2 bash -c "seq 5080 | awk '"'"'{print 12345678901234567890}'"'"' "'
<works>
```

**Comment by sebastian-philipp** [comment]
close, as we have a workaround, now?

**Comment by tserong** [comment]
Fine by me

**Comment by sebastian-philipp** [comment]
workaround implements: https://github.com/saltstack/salt/issues/50851

**Comment by ricardoasmarques** [comment]
Agree
